src/lib/proof-system/workers.ts

```diff
@@ -15,3 +15,68 @@ const workers = {
 const setNumberOfWorkers = (numWorkers: number) => {
   workers.numWorkers = numWorkers;
 };
+
+type ThreadPoolState =
+  | { type: 'none' }
+  | { type: 'initializing'; initPromise: Promise<void> }
+  | { type: 'running' }
+  | { type: 'exiting'; exitPromise: Promise<void> };
+
+function WithThreadPool({
```

Comment:
the added logic here comes from the bindings, see https://github.com/o1-labs/o1js-bindings/pull/318